module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -693,6 +808,31 @@ func headAccessibleFromTail(t *testing.T, headSliceIndex EIndex, tailSliceIndex
 		_, ok := seen[index]
 		require.False(t, ok, "duplicate identifiers found")
 
-		index = pool.poolEntities[index].node.prev.getSliceIndex()
+		index = pool.poolEntities[index].node.prev
+	}
+}
+
+// checkEachEntityIsInFreeOrUsedState checks if each entity in the pool belongs exactly to one of the state lists.
+func checkEachEntityIsInFreeOrUsedState(t *testing.T, pool *Pool) {
+	pool_capacity := len(pool.poolEntities)
+	// check size
+	require.Equal(t, int(pool.free.size+pool.used.size), pool_capacity, "Pool capacity is not equal to the sum of used and free sizes")
+	// check elelments
+	nodesInFree := discoverEntitiesBelongingToStateList(t, pool, &pool.free)
+	nodesInUsed := discoverEntitiesBelongingToStateList(t, pool, &pool.used)
+	for i := 0; i < pool_capacity; i++ {
+		require.False(t, !nodesInFree[i] && !nodesInUsed[i], "Node is not in any state list")
+		require.False(t, nodesInFree[i] && nodesInUsed[i], "Node is in two state lists at the same time")
+	}
+}
+
+// discoverEntitiesBelongingToStateList discovers all entities in the pool that belong to the given list.
+func discoverEntitiesBelongingToStateList(t *testing.T, pool *Pool, s *state) []bool {
```

Comment:
Please also refactor this using the `StateType` instead of passing a reference to the state.

Reply:
done, please take a look